pkg/ipam/ipam.go

```diff
@@ -193,89 +193,143 @@ func (ipam *IPAM) AddOrUpdateSubnet(name, cidrStr, gw string, excludeIps []strin
-			subnet.V4Free = ips.Separate(subnet.V4Reserved)
-			subnet.V4Available = subnet.V4Free.Clone()
-			subnet.V4Using = subnet.V4Using.Intersect(ips)
-			subnet.V4Gw = v4Gw
-
-			pool := subnet.IPPools[""]
-			pool.V4IPs = ips
-			pool.V4Free = subnet.V4Free.Clone()
-			pool.V4Reserved = subnet.V4Reserved.Clone()
-			pool.V4Released = NewEmptyIPRangeList()
-			pool.V4Using = subnet.V4Using.Clone()
-
-			for name, p := range subnet.IPPools {
-				if name == "" {
-					continue
+		if protocol == kubeovnv1.ProtocolDual || protocol == kubeovnv1.ProtocolIPv4 {
+			if subnet.V4CIDR.String() != v4cidrStr || subnet.V4Gw != v4Gw {
+				_, cidr, _ := net.ParseCIDR(v4cidrStr)
+				subnet.V4CIDR = cidr
+				subnet.V4Reserved = v4Reserved
+				firstIP, _ := util.FirstIP(v4cidrStr)
+				lastIP, _ := util.LastIP(v4cidrStr)
+				ips, _ := NewIPRangeListFrom(fmt.Sprintf("%s..%s", firstIP, lastIP))
+				subnet.V4Free = ips.Separate(subnet.V4Reserved)
+				subnet.V4Available = subnet.V4Free.Clone()
+				subnet.V4Using = subnet.V4Using.Intersect(ips)
+				subnet.V4Gw = v4Gw
+
+				pool := subnet.IPPools[""]
+				pool.V4IPs = ips
+				pool.V4Free = subnet.V4Free.Clone()
+				pool.V4Reserved = subnet.V4Reserved.Clone()
+				pool.V4Released = NewEmptyIPRangeList()
+				pool.V4Using = subnet.V4Using.Clone()
+
+				for name, p := range subnet.IPPools {
+					if name == "" {
+						continue
+					}
+					p.V4Free = ips.Intersect(p.V4IPs)
+					p.V4Reserved = subnet.V4Reserved.Intersect(p.V4IPs)
+					p.V4Available = p.V4Free.Clone()
+					p.V4Released = NewEmptyIPRangeList()
+					pool.V4Free = pool.V4Free.Separate(p.V4IPs)
+					pool.V4Reserved = pool.V4Reserved.Separate(p.V4Reserved)
+				}
+				pool.V4Available = pool.V4Free.Clone()
+
+				for nicName, ip := range subnet.V4NicToIP {
+					if !ips.Contains(ip) {
+						podName := subnet.V4IPToPod[ip.String()]
+						klog.Errorf("%s address %s not in subnet %s new cidr %s", podName, ip.String(), name, cidrStr)
+						delete(subnet.V4NicToIP, nicName)
+						delete(subnet.V4IPToPod, ip.String())
+						continue
+					}
 				}
-				p.V4Free = ips.Intersect(p.V4IPs)
-				p.V4Reserved = subnet.V4Reserved.Intersect(p.V4IPs)
-				p.V4Available = p.V4Free.Clone()
-				p.V4Released = NewEmptyIPRangeList()
-				pool.V4Free = pool.V4Free.Separate(p.V4IPs)
-				pool.V4Reserved = p.V4Reserved.Separate(p.V4Reserved)
 			}
-			pool.V4Available = pool.V4Free.Clone()
-
-			for nicName, ip := range subnet.V4NicToIP {
-				if !ips.Contains(ip) {
-					podName := subnet.V4IPToPod[ip.String()]
-					klog.Errorf("%s address %s not in subnet %s new cidr %s", podName, ip.String(), name, cidrStr)
-					delete(subnet.V4NicToIP, nicName)
-					delete(subnet.V4IPToPod, ip.String())
-					continue
+			if !subnet.V4Reserved.Equal(v4Reserved) {
+				newAddV4Reserved := v4Reserved.Separate(subnet.V4Reserved)
+				// only support add reserved ip,because it will cause the ip repeated sometime when you add reserved ip and delete after sometime.
```

Comment:
```suggestion
				// only support add reserved ip, because it will cause the ip repeated sometime when you add reserved ip and delete after sometime.
```